core/services/workflows/metering/metering.go

```diff
@@ -111,24 +112,64 @@ type Report struct {
 	lggr    logger.Logger
 
 	// internal state
+	ready bool
 	mu    sync.RWMutex
-	steps map[ReportStepRef]ReportStep
+	steps map[string]ReportStep
 }
 
 func NewReport(accountID, workflowID, workflowExecutionID string, lggr logger.Logger) *Report {
-	logger := lggr.Named("Metering")
-
+	sugaredLggr := logger.Sugared(lggr).Named("WorkflowEngine").With("workflowID", workflowID, "workflowExecutionID", workflowExecutionID)
 	return &Report{
 		accountID:           accountID,
 		workflowID:          workflowID,
 		workflowExecutionID: workflowExecutionID,
-		balance:             NewBalanceStore(0, map[string]decimal.Decimal{}, logger),
-		lggr:                logger,
-		steps:               make(map[ReportStepRef]ReportStep),
-		refCount:            make(map[ReportStepRef]uint64),
+
+		balance: NewBalanceStore(0, map[string]decimal.Decimal{}, sugaredLggr),
+		lggr:    sugaredLggr,
+
+		ready: false,
+		steps: make(map[string]ReportStep),
 	}
 }
 
+func (r *Report) Initialize(ctx context.Context) error {
+	if r.client == nil {
+		// TODO: more robust check of billing service health
+		return errors.New("no billing client configured")
+	}
+
+	// TODO: get rate card from billing service
+	rateCard := map[string]decimal.Decimal{}
+
+	balanceStore := NewBalanceStore(0, rateCard, r.lggr)
+
+	// If there is no credit limit defined in the workflow, then open an empty reservation
+	// TODO: consume user defined workflow execution limit
+	req := billing.ReserveCreditsRequest{
+		AccountId:           r.accountID,
+		WorkflowId:          r.workflowID,
+		WorkflowExecutionId: r.workflowExecutionID,
+		Credits:             []*billing.AccountCreditsInput{}, // TODO: send the credit balance, not resource types
+	}
+
+	resp, err := r.client.ReserveCredits(ctx, &req)
+	// If there is an error communicating with the billing service, fail openly
+	if err != nil {
+		// TODO: track failure
+		balanceStore.AllowNegative()
+		r.lggr.Warnf("failed to reserve credits: %s", err)
+	} else {
+		success := resp.GetSuccess()
+		// TODO: once response contains balance set using balanceStore.Add
+		if !success {
+			return errors.New("insufficient balance funding")
```

Comment:
To be thorough, can we export this error as a type? `ErrInsufficientBalance` in the case we need to do a type check on it later.